controllers/ca_controller.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/kubewarden/kubewarden-controller/internal/pkg/admission"
+	"github.com/kubewarden/kubewarden-controller/internal/pkg/admissionregistration"
+	"github.com/kubewarden/kubewarden-controller/internal/pkg/constants"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// This controller reconcile the root ca used by the Kubewarden stack
+type KubewardenCAReconciler struct {
+	client.Client
+	Log        logr.Logger
+	Scheme     *runtime.Scheme
+	Reconciler admission.Reconciler
+}
+
+func (r *KubewardenCAReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	CASecret, initialized, err := r.Reconciler.FetchOrInitializeRootCASecret(ctx, admissionregistration.GenerateCA, admissionregistration.PemEncodeCertificate)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to create root CA")
```

Comment:
Technically we are not creating the CA inside of the etcd store yet. A better error message could be: `failed to fetch root CA`

Reply:
Actually it can try to create the root CA. For example, if the secret is deleted. This controller we recreated. Thus, the error could happen during the secret creation. Let me change the error message to be more generic and leave the original error to give more context about the actual errors.